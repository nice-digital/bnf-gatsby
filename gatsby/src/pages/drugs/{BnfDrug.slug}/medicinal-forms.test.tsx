import { within, screen, render, waitFor } from "@testing-library/react";

import MedicinalFormsPage, {
	query,
	type MedicinalFormsPageProps,
} from "./medicinal-forms";

const drug: MedicinalFormsPageProps["data"]["bnfDrug"] = {
	title: "Anti-D (Rh<sub>0</sub>) immunoglobulin",
	slug: "anti-d-rh0-immunoglobulin",
	medicinalForms: {
		initialStatement: "",
		specialOrderManufacturersStatement: "",
		medicinalForms: [
			{
				form: "Tablets",
				slug: "tablets",
				cautionaryAndAdvisoryLabels: [],
				electolytes: null,
				excipients: null,
				preps: [],
				cautionaryAndAdvisoryLabels: [
					{
						label: {
							number: 3,
							description: "<p>Description</p>",
							englishRecommendation: "<p>English recommendation</p>",
							welshRecommendation: "<p>Welsh recommendation</p>",
						},
						additionalInfo: "test info",
					},
					{
						label: {
							number: 5,
							description: "<p>Another description</p>",
							englishRecommendation: "<p>Another English recommendation</p>",
							welshRecommendation: "<p>Another Welsh recommendation</p>",
						},
					},
				],
			},
			{
				form: "Powder",
				slug: "powder",
				cautionaryAndAdvisoryLabels: [],
				electolytes: null,
				excipients: null,
				preps: [],
			},
		],
	},
};

const dataProp: MedicinalFormsPageProps["data"] = {
	bnfDrug: drug,
};

describe("MedicinalFormsPage", () => {
	it("should match snapshot for graphql query", () => {
		expect(query).toMatchSnapshot();
	});

	describe("SEO", () => {
		it("should strip HTML tags for page title", async () => {
			render(<MedicinalFormsPage data={dataProp} />);

			await waitFor(() => {
				expect(document.title).toStartWith(
					"Medicinal forms | Anti-D (Rh0) immunoglobulin | Drugs |"
				);
			});
		});

		it.todo("Meta description (see BNF-1215");
	});

	describe("Breadcrumbs", () => {
		it.each([
			["NICE", "https://www.nice.org.uk/"],
			["BNF", "/"],
			["Drugs", "/drugs/"],
			["Anti-D (Rh0) immunoglobulin", "/drugs/anti-d-rh0-immunoglobulin/"],
		])(
			"should render default '(%s)' breadcrumb",
			(breadcrumbText, expectedHref) => {
				render(<MedicinalFormsPage data={dataProp} />);

				const breadcrumbNav = screen.getByRole("navigation", {
					name: "Breadcrumbs",
				});
				const breadcrumb = within(breadcrumbNav).getByRole("link", {
					name: breadcrumbText,
				});

				expect(breadcrumb).toHaveAttribute("href", expectedHref);
			}
		);

		it("should render current page breadcrumb without link", () => {
			render(<MedicinalFormsPage data={dataProp} />);

			const breadcrumbNav = screen.getByRole("navigation", {
				name: "Breadcrumbs",
			});
			const currentPageCrumb =
				within(breadcrumbNav).getByText("Medicinal forms");
			expect(currentPageCrumb).toBeInTheDocument();
			expect(currentPageCrumb.tagName).toBe("SPAN");
		});
	});

	describe("Body", () => {
		it("should create a section and heading for each medicinal form", () => {
			render(<MedicinalFormsPage data={dataProp} />);

			expect(screen.getAllByRole("region")).toHaveLength(2);
			expect(
				screen.getAllByRole("heading", { level: 2, name: /Tablets|Powder/ })
			).toHaveLength(2);
		});

		it("should create a labelled section", () => {
			render(<MedicinalFormsPage data={dataProp} />);

			expect(
				screen.getByRole("region", { name: "Tablets" })
			).toBeInTheDocument();
		});

		it("should show cautionary labels when they have been supplied", () => {
			render(<MedicinalFormsPage data={dataProp} />);

			const labelAccordion = screen.getByRole("group");
			expect(labelAccordion).toHaveClass("labelAccordion");

			const labelList = within(labelAccordion).getByRole("list");
			const numberOfLabels =
				dataProp.bnfDrug.medicinalForms.medicinalForms[0]
					.cautionaryAndAdvisoryLabels?.length || -1;
			// eslint-disable-next-line testing-library/no-node-access
			expect(labelList.children).toHaveLength(numberOfLabels);
		});
	});
});
