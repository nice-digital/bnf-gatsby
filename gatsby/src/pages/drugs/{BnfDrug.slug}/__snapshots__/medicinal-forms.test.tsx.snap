// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MedicinalFormsPage Body should match snapshot for page body 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="https://www.nice.org.uk/"
        >
          NICE
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/"
        >
          BNF
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/drugs/"
        >
          Drugs
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/drugs/anti-d-rh0-immunoglobulin/"
        >
          Anti-D (Rh0) immunoglobulin
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Medicinal forms
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "https://www.nice.org.uk/",
        "name": "NICE"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/",
        "name": "BNF"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/drugs/",
        "name": "Drugs"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "@id": "/drugs/anti-d-rh0-immunoglobulin/",
        "name": "Anti-D (Rh0) immunoglobulin"
      }
    },
    {
      "@type": "ListItem",
      "position": 5,
      "item": {
        "name": "Medicinal forms"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
    id="content-start"
  >
    <h1
      class="page-header__heading"
    >
      <span
        class="page-header__pre-heading"
      >
        <span>
          Anti-D (Rh
          <sub>
            0
          </sub>
          ) immunoglobulin 
        </span>
      </span>
      <span>
        Medicinal forms
      </span>
    </h1>
    <p
      class="page-header__cta"
    >
      <a
        href="/drugs/anti-d-rh0-immunoglobulin/"
      >
        View 
        anti-D (Rh0) immunoglobulin
         drug monograph
      </a>
    </p>
  </div>
  <div
    class="grid grid--loose"
    data-testid="body"
  >
    <div
      data-g="12 md:12 lg:12"
    >
      <div
        class="grid grid--rev grid--loose"
      >
        <div
          data-g="12 lg:3"
        >
          <div
            class="alert alert--info"
          >
            <p>
              NICE does not sell any drugs, medicines or pharmaceutical products.
            </p>
          </div>
        </div>
        <div
          class="body"
          data-g="12 lg:9"
        >
          <nav
            aria-labelledby="navigate-to-section"
            class="nav sectionNav"
          >
            <h2
              class="heading"
              id="navigate-to-section"
            >
              Navigate to section
            </h2>
            <ol
              aria-label="Jump links to sections on this page"
              class="linkList"
            >
              <li>
                <a
                  href="#tablets"
                >
                  Tablets
                </a>
              </li>
              <li>
                <a
                  href="#powder"
                >
                  Powder
                </a>
              </li>
            </ol>
          </nav>
          <p>
            Test initial statement
          </p>
          <p>
            Test special order statement
          </p>
          <section
            aria-labelledby="tablets"
            class="form"
          >
            <h2
              id="tablets"
            >
              Tablets
            </h2>
            <p>
              All products
            </p>
            <details
              class="details labelAccordion warningTheme"
            >
              <summary
                class="summary"
                data-tracking="Show"
              >
                <span
                  class="toggleLabel label"
                >
                  <svg
                    aria-hidden="true"
                    class="icon"
                    fill="none"
                    focusable="false"
                    height="1em"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M478.152 174c0 10.091-3.525 18.661-10.576 25.712l-186 186c-7.243 7.243-15.909 10.864-26 10.864-10.283 0-18.853-3.621-25.712-10.864l-186-186C36.621 192.853 33 184.283 33 174c0-10.091 3.621-18.757 10.864-26L65 126.576C72.424 119.525 81.09 116 91 116c10.091 0 18.661 3.525 25.712 10.576L255.576 265.44 394.44 126.576c7.051-7.051 15.621-10.576 25.712-10.576 9.909 0 18.576 3.525 26 10.576L467.576 148c7.051 7.424 10.576 16.091 10.576 26z"
                      fill="currentColor"
                    />
                  </svg>
                  Show
                </span>
                 
                <h3
                  class="labelAccordionHeading"
                >
                  <svg
                    aria-hidden="true"
                    class="warningIcon"
                    fill="none"
                    focusable="false"
                    height="1em"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M429.067 456.4H82.342c-20.328 0-36.703-8.151-44.91-22.369-8.195-14.205-7.081-32.449 3.097-50.05L213.89 83.695C224.07 66.093 239.302 56 255.718 56c16.417 0 31.649 10.107 41.814 27.709L470.88 383.995c10.178 17.602 11.306 35.845 3.097 50.05-8.208 14.204-24.568 22.355-44.91 22.355zM255.718 92.445c-2.384 0-6.467 2.927-10.249 9.493L72.106 402.21c-3.783 6.553-4.282 11.563-3.097 13.619 1.185 2.056 5.767 4.126 13.347 4.126h346.739c7.566 0 12.148-2.084 13.348-4.14 1.199-2.056.685-7.052-3.098-13.619L265.983 101.91c-3.783-6.552-7.88-9.479-10.25-9.479l-.015.014zm-21.641 202.898c.956 9.95 2.57 17.274 4.939 22.385 2.97 6.438 8.48 9.835 15.946 9.835 7.309 0 12.848-3.44 16.046-9.964 2.598-5.31 4.24-12.534 4.982-22.013l6.424-73.633a200.589 200.589 0 001.071-20.571c0-12.078-1.585-21.2-4.854-27.923-2.627-5.396-8.566-11.821-21.87-11.821-8.551 0-15.589 2.898-20.928 8.594-5.254 5.611-7.909 13.32-7.909 22.927 0 6.195.457 16.417 1.356 30.407l4.797 71.777zm1.884 56.004c5.325-5.325 11.906-8.038 19.601-8.038 7.651 0 14.275 2.684 19.7 7.98 5.467 5.339 8.223 11.906 8.223 19.501 0 8.522-2.856 15.36-8.466 20.357-5.453 4.854-11.891 7.295-19.157 7.295-7.352 0-13.862-2.413-19.344-7.181-5.696-4.94-8.594-11.835-8.594-20.457 0-7.58 2.712-14.133 8.037-19.457z"
                      fill="currentColor"
                      fill-rule="evenodd"
                    />
                  </svg>
                  Cautionary and advisory labels
                </h3>
              </summary>
              <ul
                class="labelList"
              >
                <li
                  class="label"
                >
                  <h4
                    class="labelHeading"
                  >
                    Label 
                    3
                     
                    (test qualifier)
                  </h4>
                  <p>
                    &lt;p&gt;English recommendation&lt;/p&gt;
                  </p>
                  <p
                    class="welsh"
                    lang="cy"
                  >
                    &lt;p&gt;Welsh recommendation&lt;/p&gt;
                  </p>
                </li>
                <li
                  class="label"
                >
                  <h4
                    class="labelHeading"
                  >
                    Label 
                    5
                     
                  </h4>
                  <p>
                    &lt;p&gt;Another English recommendation&lt;/p&gt;
                  </p>
                  <p
                    class="welsh"
                    lang="cy"
                  >
                    &lt;p&gt;Another Welsh recommendation&lt;/p&gt;
                  </p>
                </li>
              </ul>
            </details>
            <h3>
              Electrolytes
            </h3>
            <p>
              May contain sodium.
            </p>
            <h3>
              Excipients
            </h3>
            <p>
              May contain alcohol, disodium edetate, polysorbates.
            </p>
          </section>
          <section
            aria-labelledby="powder"
            class="form"
          >
            <h2
              id="powder"
            >
              Powder
            </h2>
            <p>
              All products
            </p>
          </section>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`MedicinalFormsPage should match snapshot for graphql query 1`] = `
"
	query ($id: String) {
		bnfDrug(id: { eq: $id }) {
			title
			slug
			medicinalForms {
				initialStatement
				specialOrderManufacturersStatement
				medicinalForms {
					form
					slug
					electrolytes
					excipients
					preps {
						...FullPrep
					}
					cautionaryAndAdvisoryLabels {
						label {
							number
							englishRecommendation
							welshRecommendation
						}
						qualifier
					}
				}
			}
		}
	}
"
`;
